In [52]:
### Importing Libraries, incl. Data Viz
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [53]:
### Create datapath to Final Project Folder - Data
datapath = r'/Users/gabrielconfrey/Desktop/Data Analysis/Career Foundry/Achievement 6/Final Project/Data'

In [54]:
### Importing data set that was cleaned and merged in Excel, via csv format
df = pd.read_csv(os.path.join(datapath, '1. Original', 'MAHousePrices.csv'))

### Need to check the import worked and that the csv came through properly formatted

In [55]:
df.columns

Index(['Town', 'Year', 'Town/Year', 'Value', 'Type', 'County', 'Population',
       '2011 % Homeowners', '2019 Per Capita Income', '2019 # Households',
       '2019 Median Household Income', '2019 Median Family Income'],
      dtype='object')

In [56]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30906 entries, 0 to 30905
Data columns (total 12 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Town                          30906 non-null  object 
 1   Year                          30906 non-null  int64  
 2   Town/Year                     30906 non-null  object 
 3   Value                         30887 non-null  object 
 4   Type                          30906 non-null  object 
 5   County                        29659 non-null  object 
 6   Population                    751 non-null    float64
 7   2011 % Homeowners             29659 non-null  float64
 8   2019 Per Capita Income        25135 non-null  object 
 9   2019 # Households             25135 non-null  float64
 10  2019 Median Household Income  25135 non-null  object 
 11  2019 Median Family Income     25043 non-null  object 
dtypes: float64(3), int64(1), object(8)
memory usage: 2.8+ MB


In [57]:
df.describe()

Year     Population  2011 % Homeowners  2019 # Households
count  30906.000000     751.000000       29659.000000       25135.000000
mean    2014.123762   30213.922770          78.844567        9414.566262
std        4.878759   57727.146122          13.312510       18041.026517
min     2006.000000    2893.000000          34.000000          60.000000
25%     2010.000000   10535.000000          72.000000        1544.000000
50%     2014.000000   18473.000000          82.000000        4968.000000
75%     2018.000000   32931.500000          88.000000       11461.000000
max     2022.000000  692600.000000          99.000000      249414.000000

In [58]:
df.shape

(30906, 12)

In [59]:
df.head(5)

Town  Year      Town/Year        Value         Type    County  \
0  Abington  2021  Abington 2021  $500,000.00  house_price  Plymouth   
1  Abington  2020  Abington 2020  $432,750.00  house_price  Plymouth   
2  Abington  2019  Abington 2019  $385,000.00  house_price  Plymouth   
3  Abington  2018  Abington 2018  $395,000.00  house_price  Plymouth   
4  Abington  2016  Abington 2016  $345,000.00  house_price  Plymouth   

   Population  2011 % Homeowners 2019 Per Capita Income  2019 # Households  \
0         NaN               73.0                $34,567             5798.0   
1         NaN               73.0                $34,567             5798.0   
2     16668.0               73.0                $34,567             5798.0   
3     16532.0               73.0                $34,567             5798.0   
4     16335.0               73.0                $34,567             5798.0   

  2019 Median Household Income 2019 Median Family Income  
0                      $78,395                   $92,452  
1                      $78,395                   $92,452  
2                      $78,395                   $92,452  
3                      $78,395                   $92,452  
4                      $78,395                   $92,452

In [60]:
### I want to know how many distinct Towns there are.
df['Town'].nunique()

367

In [61]:
### Because I input values in the 2019 Per Capita, Households, Household Income, and Family Income columns
### for each unique Town, I want to run value counts to see if the unique values in those columns matches Town.
df['2019 Per Capita Income'].nunique()

284

In [62]:
df['2019 # Households'].nunique()

286

In [63]:
df['2019 Median Household Income'].nunique()

282

In [64]:
df['2019 Median Family Income'].nunique()

282

### It's evident that I did not have data for each unique town, which will make comparisons unproductive.
### Next step is to identify which Towns do not have values in those columns and delete their rows from the data.
### Given 367 Towns, it's expected that 85 of them will not have all the data (2019 Median Family Income = 282).

In [65]:
### Find NaN values in '2019 Median Family Income'
df['2019 Median Family Income'].isna()

0        False
1        False
2        False
3        False
4        False
         ...  
30901    False
30902    False
30903    False
30904    False
30905    False
Name: 2019 Median Family Income, Length: 30906, dtype: bool

In [66]:
df['2019 Median Family Income'].isna().count()

30906

In [67]:
dfNaN = (df[df['2019 Median Family Income'].isnull()])

In [68]:
dfNaN.head()

Town  Year      Town/Year        Value         Type County  \
14  Amesbury  2021  Amesbury 2021  $525,000.00  house_price  Essex   
15  Amesbury  2020  Amesbury 2020  $463,000.00  house_price  Essex   
16  Amesbury  2019  Amesbury 2019  $415,000.00  house_price  Essex   
17  Amesbury  2018  Amesbury 2018  $390,950.00  house_price  Essex   
18  Amesbury  2016  Amesbury 2016  $330,925.00  house_price  Essex   

    Population  2011 % Homeowners 2019 Per Capita Income  2019 # Households  \
14         NaN               70.0                    NaN                NaN   
15         NaN               70.0                    NaN                NaN   
16     17532.0               70.0                    NaN                NaN   
17     17510.0               70.0                    NaN                NaN   
18     17327.0               70.0                    NaN                NaN   

   2019 Median Household Income 2019 Median Family Income  
14                          NaN                       NaN  
15                          NaN                       NaN  
16                          NaN                       NaN  
17                          NaN                       NaN  
18                          NaN                       NaN

In [69]:
dfNaN['Town'].nunique()

77

### By subsetting the null values, and counting the unique Towns in that subset, it seems there are 77 towns with missing data in the Family Income column. This number is relatively close to the expected.
### Because these Towns don't have the necessary information for my analysis, I could either find the information via another source, or delete the Towns with NaN in that column.
### I will do the latter.

In [72]:
dfNaN["Town"].unique()

array(['Amesbury', 'Bellingham', 'Braintree', 'Cambridge', 'Essex',
       'Hull', 'Manchester-by-the-Sea', 'Melrose', 'Middleboro', 'Millis',
       'Nahant', 'Newbury', 'Norfolk', 'North Attleboro', 'Pembroke',
       'Pepperell', 'Saugus', 'Southboro', 'Townsend', 'Watertown',
       'Weston', 'Barnstable Town', 'Chatham', 'Orleans', 'Province town',
       'Truro', 'Wellfleet', 'Alford', 'Egremont', 'Mount Washington',
       'New Ashford', 'Richmond', 'Tyringham', 'Williams town',
       'Free town', 'Aquinnah', 'Chilmark', 'Edgar town', 'George town',
       'Buckland', 'Leverett', 'East Longmeadow', 'Longmeadow', 'Tolland',
       'Belcher town', 'Hadley', 'Plainfield', 'South Hadley',
       'Water city', 'Marion', 'Auburn', 'Berlin', 'Bolton', 'Boylston',
       'Grafton', 'Holden', 'Hopedale', 'Mendon', 'Northborough',
       'Phillipston', 'Princeton', 'Sterling', 'Sutton', 'Upton',
       'Uxbridge', 'Westborough', 'Amesbury Town', 'Greenfield Town',
       'Agawam Town', '

In [78]:
df.dropna(subset = ['2019 Median Family Income'], inplace=True)

In [85]:
df['2019 Median Family Income'].isna().sum()

0

### I have successfully removed the null values from the 2019 Median Family Income column.

In [86]:
df.shape

(25043, 12)

In [87]:
df['Town'].nunique()

290

### There are now 290 unique Towns, which is 77 less than before. This matches our expectation.

In [88]:
### Last step is to delete unnecessary columns: "Town/Date".
df.drop(columns = ['Town/Year'])

Town  Year        Value         Type     County  Population  \
0       Abington  2021  $500,000.00  house_price   Plymouth         NaN   
1       Abington  2020  $432,750.00  house_price   Plymouth         NaN   
2       Abington  2019  $385,000.00  house_price   Plymouth     16668.0   
3       Abington  2018  $395,000.00  house_price   Plymouth     16532.0   
4       Abington  2016  $345,000.00  house_price   Plymouth     16335.0   
...          ...   ...          ...          ...        ...         ...   
30901  Worcester  2022    $1,103.00     Br0_rent  Worcester         NaN   
30902  Worcester  2022    $1,162.00     Br1_rent  Worcester         NaN   
30903  Worcester  2022    $1,491.00     Br2_rent  Worcester         NaN   
30904  Worcester  2022    $1,843.00     Br3_rent  Worcester         NaN   
30905  Worcester  2022    $2,022.00     Br4_rent  Worcester         NaN   

       2011 % Homeowners 2019 Per Capita Income  2019 # Households  \
0                   73.0                $34,567             5798.0   
1                   73.0                $34,567             5798.0   
2                   73.0                $34,567             5798.0   
3                   73.0                $34,567             5798.0   
4                   73.0                $34,567             5798.0   
...                  ...                    ...                ...   
30901               49.0                $24,330            68850.0   
30902               49.0                $24,330            68850.0   
30903               49.0                $24,330            68850.0   
30904               49.0                $24,330            68850.0   
30905               49.0                $24,330            68850.0   

      2019 Median Household Income 2019 Median Family Income  
0                          $78,395                   $92,452  
1                          $78,395                   $92,452  
2                          $78,395                   $92,452  
3                          $78,395                   $92,452  
4                          $78,395                   $92,452  
...                            ...                       ...  
30901                      $45,932                   $57,704  
30902                      $45,932                   $57,704  
30903                      $45,932                   $57,704  
30904                      $45,932                   $57,704  
30905                      $45,932                   $57,704  

[25043 rows x 11 columns]

In [89]:
### One last value_count for Type though:
df['Type'].value_counts()

Type
Br0_rent       4823
Br1_rent       4823
Br2_rent       4823
Br3_rent       4823
Br4_rent       4823
house_price     928
Name: count, dtype: int64

### Because I'm only interested in houses or rentals with 2 or more bedrooms, I can also delete rows where Type = 'Br0_rent'

In [90]:
### Deleting rows where 'Type' = Br0_rent
df.drop(df.loc[df['Type']== 'Br0_rent'].index, inplace=True)

In [91]:
df['Type'].value_counts()

Type
Br1_rent       4823
Br2_rent       4823
Br3_rent       4823
Br4_rent       4823
house_price     928
Name: count, dtype: int64

In [92]:
### Deleting rows where 'Type' = Br1_rent
df.drop(df.loc[df['Type']== 'Br1_rent'].index, inplace=True)

In [93]:
df['Type'].value_counts()

Type
Br2_rent       4823
Br3_rent       4823
Br4_rent       4823
house_price     928
Name: count, dtype: int64

In [99]:
df['Type'].describe()

count        15397
unique           4
top       Br2_rent
freq          4823
Name: Type, dtype: object

In [100]:
df.shape

(15397, 12)

# After cleaning, we have 15,397 rows of data to analyze, with the necessary column values to answer the following research questions:

### The area needs to be
1) affordable, defined as median home price <$400,000, and the median family income <=$100,000.
2) in a residential area, defined as being 75% home-owned
3) within a 1 hr drive (30 mi) of Boston or (50 mi) Manchester, VT.


In [107]:
### Exporting dataset into Cleaned as pkl
df.to_pickle(os.path.join(datapath, '2. Cleaned', 'homedataclean.pkl'))